presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java

```diff
@@ -191,7 +205,62 @@ private QualifiedName generateTemporaryTableName(Optional<QualifiedName> origina
         else {
             zeroRowQuery = new Query(query.getWith(), query.getQueryBody(), Optional.empty(), Optional.of("0"));
         }
-        return prestoAction.execute(zeroRowQuery, REWRITE, ResultSetConverter.DEFAULT).getColumnNames();
+        return prestoAction.execute(zeroRowQuery, REWRITE, ResultSetConverter.DEFAULT).getMetadata();
+    }
+
+    private Query rewriteNonStorableColumns(Query query, ResultSetMetaData metadata)
+    {
+        // Skip if all columns are storable
+        List<TypeSignature> columnTypes = getColumnTypes(metadata);
+        if (columnTypes.stream().noneMatch(type -> getColumnTypeRewrite(type).isPresent())) {
+            return query;
+        }
+
+        // Cannot handle SELECT query with top-level SetOperation
+        if (!(query.getQueryBody() instanceof QuerySpecification)) {
+            return query;
+        }
+
+        QuerySpecification querySpecification = (QuerySpecification) query.getQueryBody();
+        List<SelectItem> selectItems = querySpecification.getSelect().getSelectItems();
+        // Cannot handle SELECT *
+        if (selectItems.stream().anyMatch(AllColumns.class::isInstance)) {
+            return query;
+        }
+
+        List<SelectItem> newItems = new ArrayList<>();
+        checkState(selectItems.size() == columnTypes.size(), "SelectItem count (%s) mismatches column count (%s)", selectItems.size(), columnTypes.size());
+        for (int i = 0; i < selectItems.size(); i++) {
+            SingleColumn singleColumn = (SingleColumn) selectItems.get(i);
+            Optional<TypeSignature> columnTypeRewrite = getColumnTypeRewrite(columnTypes.get(i));
+            if (columnTypeRewrite.isPresent()) {
+                newItems.add(new SingleColumn(new Cast(singleColumn.getExpression(), columnTypeRewrite.get().toString()), singleColumn.getAlias()));
+            }
+            else {
+                newItems.add(singleColumn);
+            }
+        }
+
+        return new Query(
+                query.getWith(),
+                new QuerySpecification(
+                        new Select(querySpecification.getSelect().isDistinct(), newItems),
+                        querySpecification.getFrom(),
+                        querySpecification.getWhere(),
+                        querySpecification.getGroupBy(),
+                        querySpecification.getHaving(),
+                        querySpecification.getOrderBy(),
+                        querySpecification.getLimit()),
+                query.getOrderBy(),
+                query.getLimit());
+    }
+
+    private static Optional<TypeSignature> getColumnTypeRewrite(TypeSignature type)
```

Comment:
This logic is Hive-specific. Does Verifier support only Hive connector?

Reply:
This is the assumption so far. Verifier are unaware about what connector is associated with a given catalog, so it won't be able make decisions around that as well.